docs/_includes/arrays/mixed_shape.yaml

```diff
@@ -0,0 +1,75 @@
+id: mixed_shape_array
+name: mixed_shape_array
+title: mixed_shape_array
+imports:
+  - linkml:types
+
+classes:
+  MixedAnyShapeArray:
+    attributes:
+      array:
+        range: integer
+        array:
+          maximum_number_dimensions: false
+          dimensions:
+            - alias: max_card
+              maximum_cardinality: 5
+            - alias: min_card
+              minimum_cardinality: 2
+            - alias: range_card
+              minimum_cardinality: 2
+              maximum_cardinality: 5
+            - alias: exact_card
+              exact_cardinality: 6
+
+  MixedMaxShapeArray:
+    attributes:
+      array:
+        range: integer
+        array:
+          maximum_number_dimensions: 6
+          dimensions:
+            - alias: max_card
+              maximum_cardinality: 5
+            - alias: min_card
+              minimum_cardinality: 2
+            - alias: range_card
+              minimum_cardinality: 2
+              maximum_cardinality: 5
+            - alias: exact_card
+              exact_cardinality: 6
+
+  MixedRangeShapeArray:
+    attributes:
+      array:
+        range: integer
+        array:
+          minimum_number_dimensions: 5
+          maximum_number_dimensions: 7
+          dimensions:
+            - alias: max_card
+              maximum_cardinality: 5
+            - alias: min_card
+              minimum_cardinality: 2
+            - alias: range_card
+              minimum_cardinality: 2
+              maximum_cardinality: 5
+            - alias: exact_card
+              exact_cardinality: 6
+
+  MixedExactShapeArray:
+    attributes:
+      array:
+        range: integer
+        array:
+          exact_number_dimensions: 6
+          dimensions:
+            - alias: max_card
+              maximum_cardinality: 5
+            - alias: min_card
+              minimum_cardinality: 2
+            - alias: range_card
+              minimum_cardinality: 2
+              maximum_cardinality: 5
+            - alias: exact_card
+              exact_cardinality: 6
```

Comment:
This is minor and not totally related to this PR, but I'm trying to imagine the use case for `minimum_number_dimensions < len(dimensions)`. 

I wonder when someone would want to specify a minimum number of dimensions but not even provide an alias hinting to the user what those dimensions are supposed to represent. 

Example 1: Let's specify an array with at least 3 dimensions and only the first one has a label. We could write:

```yaml
array:
  minimum_number_dimensions: 3
  maximum_number_dimensions: False
  dimensions:
    - alias: time
```

The user must recognize that `dimensions` does not provide information on neither the 2nd and 3rd required dimensions, nor the dimensions beyond that. 

Example 2: Example 1 is equivalent to this schema:

```yaml
array:
  minimum_number_dimensions: 3
  maximum_number_dimensions: False
  dimensions:
    - alias: time
    - {}
    - {}
```

The dimensions beyond the first 3 still have no information here. You could still provide information on those dimensions by adding more list items to `dimensions`. 

Example 2 is less compact but it might be more intuitive. It looks a little ugly with the `{}` but I think we might want to encourage schema writers to provide at least some information on the required dimensions. I don't see why they could not, and if they do not want to, then they have to explicitly say there is nothing with `{}`. 

But I've looked at this too long to be a good judge of what is and is not intuitive. So I guess my question is whether we want to allow the "shortcut" schema pattern shown in Example 1. What do you think @sneakers-the-rat @cmungall ? 

(In NWB, you can define the allowed shapes without providing any information about the shapes, but it is generally recommended that you label the dimensions. In some cases, it might feel unnecessary, e.g., a 1D array of timestamps has the first dimension labeled as "times" or "num_times".)

Reply:
Assuming you mean `minimum_number_dimensions > length(dimensions)` at the top by context :)

In this case, what i was thinking about was providing a complete and predictable behavior based on the syntactic tools provided. We could make a rule like "minimum number dimensions must be equal to number of specified dimensions" but that would be an additional constraint on top of the basic rules of the syntax that would require additional documentation and special validation logic. 

The implementation burden is relatively low, and imo a simple syntax that does the things you expect it to is good to do, even if we can't think of times when it might be useful! I don't really like enforcing or nudging design patterns at the level of a syntax, that takes us out of the realm of 'neutral expressive tool that can accomodate many formats' and more into the realm of 'opinionated data format,' in this case both of those should be valid (and equivalent) - and probably should be added as test cases since they are good examples of edge cases.

Reply:
does this deserve to be in 'linkml arrays, extended docs universe', a separate issue, or keep this open on the issue for history's sake?

Reply:
I think this is fine to resolve. Thanks for the response! Being a neutral expressive tool as opposed to opinionated data format makes sense here. LinkML could have a separate layer of best practices that suggests one of the above formats over the other for clarity of those reading the raw schema. I added an issue about that here https://github.com/linkml/linkml/issues/2025